features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,263 @@
+Fundamentally, auth gear with web-hooks is a distributed system.
+When web-hook handlers have side-effects, we need to choose between guaranteeing
+consistency or availability of the system (see appendix: CAP theorem).
 
-| argument | description | require |
-| -------- | -------- | -------- | 
-| `events` | A list of gear events, such as: ["after_signup", "before_login"]. | ✓ |
-| `url` | The url of the webhook. | ✓ |
-| `async` | default is `true`. | |
-| `secret` | default is empty, if provided, it will be used as the key to generate `X-Skygear-Webhook-Signature` digest. | |
-| `timeout` | - default value of sync hook is 5 seconds, it allows to be configured up to 10 seconds.<br/>- default value of async hook is 60 seconds and up to 300 seconds.  | |
+We decided to ensure the availability of the system. To maintain consistency,
+developer should take eventual consistency into account when designing system.
 
-## versioning
+Developer should regular check the past event list for unprocessed events to
+ensure system consistency.
 
-Like skygear cloud functions, webhooks should be consider as a "moving part". So no mattter a user deploy webhooks by `skycli` or gear REST interface, a new app version should be created as well. Note that, a new app version and a new unit version will be generated when deploy by REST interface every time.
+> We considered ensuring the consistency of the system through distributed
+> transaction protocols (e.g. two-phase commit). However, its implementation
+> requires much effort from both auth gear and web-hook handlers, and introduce
+> a single point of failure (coordinator). We decided on availability instead.
 
-## Verify Request
+### Event Timing
+There are four possible delivery timing of events: sync BEFORE, async BEFORE,
+sync AFTER, async AFTER.
 
-When `secret` of a webhook is not empty, `X-Skygear-Webhook-Signature` will be added to request headers, it is the HMAC hex digest of the POST request payload. The digest is generated using the sha256 hash function and the hook `secret` as the key.
+Async BEFORE is mostly useless: the request may not be successful, and handler
+cannot affect the operation. Therefore, we do not offer async BEFORE events.
 
-## Timeout
+Sync AFTER cannot be used safely:
+- If it is not within the operation transaction, async AFTER can be used instead.
+- If it is within the operation transaction,
+  then the transaction should rollback on web-hook delivery failure,
+  otherwise async AFTER can be used instead. So:
+    - If the handler has no side-effect, sync BEFORE can be used instead
+      (e.g. custom validation)
+    - If the handler has side-effect, it is subject to distributed transaction
+      problem. We use eventual consistency so async AFTER should be used.
 
-A `SYNC` hook is considered failed if it can't response within 5 seconds, where the value of timeout is configurable of each hook (up to 10 seconds).
+Therefore, we do not offer sync AFTER events.
```

Comment:
In Skygear v1 experience, sync after is rarely useful except when some data are only available AFTER the operation is completed (let's say loginWithOAuth and we only fill in the profile afterwards)

I guess it is a bit hard to think of those cases now, but is it possible we leave the possibility if one day we really need a SYNC after, it could be done?

Reply:
As explained, I believe that a sync after event is useless and cannot be used safely. Therefore I don't think we should consider supporting sync after in future.

Reply:
@kiootic as suggested, there are cases it could fit the criteria of, say data are only populated after the action, and the hooks want to be called with the data after the action, so your assumption of sync after the event is useless and cannot be used safely could be invalid (at least it is not a logical proof that I'm very confident in)

I didn't go into the details of all possible use cases and hooks we will provide yet. So I'm suggesting if it is an unknown, maybe we want to make sure our design can cater that if it come back to us.

Maybe let me give a sample which I'm unsure if it will happen:
-- e.g. Let's say we have an API for facebook login, only after the login is done (after identities created? after_user_create?) we will have the claims from FB.

However, the user wants to avoid any Facebook login with email address ends with "@gmail.com" to create an account, it seems a perfect case with after_user_create but sync?

Reply:
Please note that as stated in the web-hook events document, the before events are delivered right before the commit into database: all possible information are already available.

In your use case, developer should use `before_identity_create` and disallow when the email in event payload is undesired.

Reply:
ok. God bless us this assumption holds in future :)